// @primer/react/dist/ActionList/Item.js
import { c } from 'react-compiler-runtime';
import React from 'react';
import { useId } from '../hooks/useId.js';
import { useSlots } from '../hooks/useSlots.js';
import { ActionListContainerContext } from './ActionListContainerContext.js';
import { Description } from './Description.js';
import { GroupContext } from './Group.js';
import { Selection } from './Selection.js';
import { TrailingVisual, LeadingVisual, VisualOrIndicator } from './Visuals.js';
import { ListContext, ItemContext } from './shared.js';
import { TrailingAction } from './TrailingAction.js';
import { ConditionalWrapper } from '../internal/components/ConditionalWrapper.js';
import { invariant } from '../utils/invariant.js';
import VisuallyHidden from '../_VisuallyHidden.js';
import classes from './ActionList.module.css.js';
import { clsx } from 'clsx';
import { fixedForwardRef } from '../utils/modern-polymorphic.js';
import { TooltipContext, Tooltip } from '../TooltipV2/Tooltip.js';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';

/**
 * Wraps button-semantic items with Tooltip, disabled when not truncated
 * For non-button-semantic items, renders children directly.
 */
const ConditionalTooltip = /*#__PURE__*/React.forwardRef(function ConditionalTooltip(t0, forwardedRef) {
  const $ = c(4);
  const {
    text,
    enabled,
    children
  } = t0;
  if (!enabled || !text) {
    return children;
  }
  const t1 = text || "";
  let t2;
  if ($[0] !== children || $[1] !== forwardedRef || $[2] !== t1) {
    t2 = /*#__PURE__*/jsx(Tooltip, {
      ref: forwardedRef,
      text: t1,
      direction: "e",
      delay: "medium",
      children: children
    });
    $[0] = children;
    $[1] = forwardedRef;
    $[2] = t1;
    $[3] = t2;
  } else {
    t2 = $[3];
  }
  return t2;
});
const SubItem = t0 => {
  const $ = c(2);
  const {
    children
  } = t0;
  let t1;
  if ($[0] !== children) {
    t1 = /*#__PURE__*/jsx(Fragment, {
      children: children
    });
    $[0] = children;
    $[1] = t1;
  } else {
    t1 = $[1];
  }
  return t1;
};
SubItem.displayName = 'ActionList.SubItem';
const ButtonItemContainerNoBox = /*#__PURE__*/React.forwardRef((t0, forwardedRef) => {
  const $ = c(9);
  let children;
  let props;
  let style;
  if ($[0] !== t0) {
    ({
      children,
      style,
      ...props
    } = t0);
    $[0] = t0;
    $[1] = children;
    $[2] = props;
    $[3] = style;
  } else {
    children = $[1];
    props = $[2];
    style = $[3];
  }
  const t1 = forwardedRef;
  let t2;
  if ($[4] !== children || $[5] !== props || $[6] !== style || $[7] !== t1) {
    t2 = /*#__PURE__*/jsx("button", {
      type: "button",
      ref: t1,
      style: style,
      ...props,
      children: children
    });
    $[4] = children;
    $[5] = props;
    $[6] = style;
    $[7] = t1;
    $[8] = t2;
  } else {
    t2 = $[8];
  }
  return t2;
});
const DivItemContainerNoBox = /*#__PURE__*/React.forwardRef((t0, forwardedRef) => {
  const $ = c(7);
  let children;
  let props;
  if ($[0] !== t0) {
    ({
      children,
      ...props
    } = t0);
    $[0] = t0;
    $[1] = children;
    $[2] = props;
  } else {
    children = $[1];
    props = $[2];
  }
  const t1 = forwardedRef;
  let t2;
  if ($[3] !== children || $[4] !== props || $[5] !== t1) {
    t2 = /*#__PURE__*/jsx("div", {
      ref: t1,
      ...props,
      children: children
    });
    $[3] = children;
    $[4] = props;
    $[5] = t1;
    $[6] = t2;
  } else {
    t2 = $[6];
  }
  return t2;
});
const baseSlots = {
  leadingVisual: LeadingVisual,
  trailingVisual: TrailingVisual,
  trailingAction: TrailingAction,
  subItem: SubItem
};
const slotsConfig = {
  ...baseSlots,
  description: Description
};

// Pre-allocated array for selectableRoles check, avoids per-render allocation
const selectableRoles = ['menuitemradio', 'menuitemcheckbox', 'option'];
const listRoleTypes = ['listbox', 'menu', 'list'];
const UnwrappedItem = (t0, forwardedRef) => {
  var _slots$trailingVisual, _slots$description$pr, _slots$description;
  const $ = c(161);
  let _PrivateItemWrapper;
  let className;
  let id;
  let inactiveText;
  let loading;
  let onSelectUser;
  let props;
  let role;
  let t1;
  let t2;
  let t3;
  let t4;
  let t5;
  if ($[0] !== t0) {
    const {
      variant: t6,
      size: t7,
      disabled: t8,
      inactiveText: t9,
      selected: t10,
      active: t11,
      onSelect: t12,
      id: t13,
      role: t14,
      loading: t15,
      _PrivateItemWrapper: t16,
      className: t17,
      groupId: _groupId,
      renderItem: _renderItem,
      handleAddItem: _handleAddItem,
      ...t18
    } = t0;
    t1 = t6;
    t2 = t7;
    t3 = t8;
    inactiveText = t9;
    t4 = t10;
    t5 = t11;
    onSelectUser = t12;
    id = t13;
    role = t14;
    loading = t15;
    _PrivateItemWrapper = t16;
    className = t17;
    props = t18;
    $[0] = t0;
    $[1] = _PrivateItemWrapper;
    $[2] = className;
    $[3] = id;
    $[4] = inactiveText;
    $[5] = loading;
    $[6] = onSelectUser;
    $[7] = props;
    $[8] = role;
    $[9] = t1;
    $[10] = t2;
    $[11] = t3;
    $[12] = t4;
    $[13] = t5;
  } else {
    _PrivateItemWrapper = $[1];
    className = $[2];
    id = $[3];
    inactiveText = $[4];
    loading = $[5];
    onSelectUser = $[6];
    props = $[7];
    role = $[8];
    t1 = $[9];
    t2 = $[10];
    t3 = $[11];
    t4 = $[12];
    t5 = $[13];
  }
  const variant = t1 === undefined ? "default" : t1;
  const size = t2 === undefined ? "medium" : t2;
  const disabled = t3 === undefined ? false : t3;
  const selected = t4 === undefined ? undefined : t4;
  const active = t5 === undefined ? false : t5;
  const [partialSlots, childrenWithoutSlots] = useSlots(props.children, slotsConfig);
  let t6;
  if ($[14] !== partialSlots) {
    t6 = {
      description: undefined,
      ...partialSlots
    };
    $[14] = partialSlots;
    $[15] = t6;
  } else {
    t6 = $[15];
  }
  const slots = t6;
  const {
    container,
    afterSelect,
    selectionAttribute,
    defaultTrailingVisual
  } = React.useContext(ActionListContainerContext);
  let t7;
  if ($[16] !== defaultTrailingVisual) {
    t7 = defaultTrailingVisual ? /*#__PURE__*/jsx(TrailingVisual, {
      children: defaultTrailingVisual
    }) : null;
    $[16] = defaultTrailingVisual;
    $[17] = t7;
  } else {
    t7 = $[17];
  }
  const wrappedDefaultTrailingVisual = t7;
  const trailingVisual = (_slots$trailingVisual = slots.trailingVisual) !== null && _slots$trailingVisual !== void 0 ? _slots$trailingVisual : wrappedDefaultTrailingVisual;
  const {
    role: listRole,
    selectionVariant: listSelectionVariant
  } = React.useContext(ListContext);
  const {
    selectionVariant: groupSelectionVariant
  } = React.useContext(GroupContext);
  const inactive = Boolean(inactiveText);
  const menuContext = container === "ActionMenu" || container === "SelectPanel" || container === "FilteredActionList";
  let t8;
  if ($[18] !== inactive || $[19] !== listRole) {
    t8 = inactive && !(listRole !== undefined && ["menu", "listbox"].includes(listRole));
    $[18] = inactive;
    $[19] = listRole;
    $[20] = t8;
  } else {
    t8 = $[20];
  }
  const showInactiveIndicator = t8;
  let t9;
  if ($[21] !== onSelectUser) {
    t9 = (event, afterSelect_0) => {
      if (typeof onSelectUser === "function") {
        onSelectUser(event);
      }
      if (event.defaultPrevented) {
        return;
      }
      if (typeof afterSelect_0 === "function") {
        afterSelect_0(event);
      }
    };
    $[21] = onSelectUser;
    $[22] = t9;
  } else {
    t9 = $[22];
  }
  const onSelect = t9;
  const selectionVariant = groupSelectionVariant ? groupSelectionVariant : listSelectionVariant;
  let inferredItemRole;
  if (container === "ActionMenu") {
    if (selectionVariant === "single") {
      inferredItemRole = "menuitemradio";
    } else {
      if (selectionVariant === "multiple") {
        inferredItemRole = "menuitemcheckbox";
      } else {
        inferredItemRole = "menuitem";
      }
    }
  } else {
    if (listRole === "listbox") {
      if (selectionVariant !== undefined && !role) {
        inferredItemRole = "option";
      }
    } else {
      if (listRole === "tablist") {
        inferredItemRole = "tab";
      }
    }
  }
  const itemRole = role || inferredItemRole;
  if (slots.trailingAction) {
    !!menuContext ? process.env.NODE_ENV !== "production" ? invariant(false, "ActionList.TrailingAction can not be used within a list with an ARIA role of \"menu\" or \"listbox\".") : invariant(false) : void 0;
  }
  let inferredSelectionAttribute;
  if (itemRole === "menuitemradio" || itemRole === "menuitemcheckbox") {
    inferredSelectionAttribute = "aria-checked";
  } else {
    if (itemRole === "option") {
      inferredSelectionAttribute = "aria-selected";
    }
  }
  const itemSelectionAttribute = selectionAttribute || inferredSelectionAttribute;
  const listItemSemantics = itemRole === "option" || itemRole === "menuitem" || itemRole === "menuitemradio" || itemRole === "menuitemcheckbox" || itemRole === "tab";
  let t10;
  if ($[23] !== inactive || $[24] !== listItemSemantics || $[25] !== listRole) {
    t10 = listRole && listRoleTypes.includes(listRole) || inactive || listItemSemantics;
    $[23] = inactive;
    $[24] = listItemSemantics;
    $[25] = listRole;
    $[26] = t10;
  } else {
    t10 = $[26];
  }
  const listSemantics = t10;
  const buttonSemantics = !listSemantics && !_PrivateItemWrapper;
  let t11;
  if ($[27] !== afterSelect || $[28] !== disabled || $[29] !== inactive || $[30] !== loading || $[31] !== onSelect) {
    t11 = event_0 => {
      if (disabled || inactive || loading) {
        return;
      }
      onSelect(event_0, afterSelect);
    };
    $[27] = afterSelect;
    $[28] = disabled;
    $[29] = inactive;
    $[30] = loading;
    $[31] = onSelect;
    $[32] = t11;
  } else {
    t11 = $[32];
  }
  const clickHandler = t11;
  let t12;
  if ($[33] !== afterSelect || $[34] !== disabled || $[35] !== inactive || $[36] !== loading || $[37] !== onSelect) {
    t12 = event_1 => {
      if (disabled || inactive || loading) {
        return;
      }
      if (event_1.key === " " || event_1.key === "Enter") {
        if (event_1.key === " ") {
          event_1.preventDefault();
          event_1.defaultPrevented = false;
        }
        onSelect(event_1, afterSelect);
      }
    };
    $[33] = afterSelect;
    $[34] = disabled;
    $[35] = inactive;
    $[36] = loading;
    $[37] = onSelect;
    $[38] = t12;
  } else {
    t12 = $[38];
  }
  const keyPressHandler = t12;
  const itemId = useId(id);
  const labelId = `${itemId}--label`;
  const inlineDescriptionId = `${itemId}--inline-description`;
  const blockDescriptionId = `${itemId}--block-description`;
  const trailingVisualId = `${itemId}--trailing-visual`;
  const inactiveWarningId = inactive && !showInactiveIndicator ? `${itemId}--warning-message` : undefined;
  const [truncatedText, setTruncatedText] = React.useState(undefined);
  const DefaultItemWrapper = listSemantics ? DivItemContainerNoBox : ButtonItemContainerNoBox;
  const ItemWrapper = _PrivateItemWrapper || DefaultItemWrapper;
  let t13;
  if ($[39] !== itemRole || $[40] !== itemSelectionAttribute) {
    t13 = itemSelectionAttribute && itemRole && selectableRoles.includes(itemRole);
    $[39] = itemRole;
    $[40] = itemSelectionAttribute;
    $[41] = t13;
  } else {
    t13 = $[41];
  }
  const includeSelectionAttribute = t13;
  const focusable = showInactiveIndicator ? true : undefined;
  const descriptionVariant = (_slots$description$pr = (_slots$description = slots.description) === null || _slots$description === void 0 ? void 0 : _slots$description.props.variant) !== null && _slots$description$pr !== void 0 ? _slots$description$pr : "inline";
  const hasTrailingVisualSlot = Boolean(slots.trailingVisual);
  const hasDescriptionSlot = Boolean(slots.description);
  let parts;
  if ($[42] !== descriptionVariant || $[43] !== hasDescriptionSlot || $[44] !== hasTrailingVisualSlot || $[45] !== inlineDescriptionId || $[46] !== labelId || $[47] !== trailingVisualId) {
    parts = [labelId];
    if (hasTrailingVisualSlot) {
      parts.push(trailingVisualId);
    }
    if (hasDescriptionSlot && descriptionVariant === "inline") {
      parts.push(inlineDescriptionId);
    }
    $[42] = descriptionVariant;
    $[43] = hasDescriptionSlot;
    $[44] = hasTrailingVisualSlot;
    $[45] = inlineDescriptionId;
    $[46] = labelId;
    $[47] = trailingVisualId;
    $[48] = parts;
  } else {
    parts = $[48];
  }
  const ariaLabelledBy = parts.join(" ");
  let parts_0;
  if ($[49] !== blockDescriptionId || $[50] !== descriptionVariant || $[51] !== hasDescriptionSlot || $[52] !== inactiveWarningId) {
    parts_0 = [];
    if (hasDescriptionSlot && descriptionVariant === "block") {
      parts_0.push(blockDescriptionId);
    }
    if (inactiveWarningId) {
      parts_0.push(inactiveWarningId);
    }
    $[49] = blockDescriptionId;
    $[50] = descriptionVariant;
    $[51] = hasDescriptionSlot;
    $[52] = inactiveWarningId;
    $[53] = parts_0;
  } else {
    parts_0 = $[53];
  }
  const ariaDescribedBy = parts_0.length > 0 ? parts_0.join(" ") : undefined;
  const t14 = !buttonSemantics ? keyPressHandler : undefined;
  const t15 = disabled ? true : undefined;
  const t16 = inactive ? true : undefined;
  const t17 = loading && !inactive ? true : undefined;
  const t18 = focusable ? undefined : 0;
  let t19;
  if ($[54] !== includeSelectionAttribute || $[55] !== itemSelectionAttribute || $[56] !== selected) {
    t19 = includeSelectionAttribute && {
      [itemSelectionAttribute]: selected
    };
    $[54] = includeSelectionAttribute;
    $[55] = itemSelectionAttribute;
    $[56] = selected;
    $[57] = t19;
  } else {
    t19 = $[57];
  }
  let t20;
  if ($[58] !== ariaDescribedBy || $[59] !== ariaLabelledBy || $[60] !== clickHandler || $[61] !== itemId || $[62] !== itemRole || $[63] !== t14 || $[64] !== t15 || $[65] !== t16 || $[66] !== t17 || $[67] !== t18 || $[68] !== t19) {
    t20 = {
      onClick: clickHandler,
      onKeyPress: t14,
      "aria-disabled": t15,
      "data-inactive": t16,
      "data-loading": t17,
      tabIndex: t18,
      "aria-labelledby": ariaLabelledBy,
      "aria-describedby": ariaDescribedBy,
      ...t19,
      role: itemRole,
      id: itemId
    };
    $[58] = ariaDescribedBy;
    $[59] = ariaLabelledBy;
    $[60] = clickHandler;
    $[61] = itemId;
    $[62] = itemRole;
    $[63] = t14;
    $[64] = t15;
    $[65] = t16;
    $[66] = t17;
    $[67] = t18;
    $[68] = t19;
    $[69] = t20;
  } else {
    t20 = $[69];
  }
  const menuItemProps = t20;
  let t21;
  if ($[70] !== _PrivateItemWrapper || $[71] !== forwardedRef || $[72] !== itemRole || $[73] !== listSemantics || $[74] !== menuItemProps || $[75] !== props) {
    t21 = _PrivateItemWrapper ? {
      role: itemRole ? "none" : undefined,
      ...props
    } : listSemantics && {
      ...menuItemProps,
      ...props,
      ref: forwardedRef
    } || {};
    $[70] = _PrivateItemWrapper;
    $[71] = forwardedRef;
    $[72] = itemRole;
    $[73] = listSemantics;
    $[74] = menuItemProps;
    $[75] = props;
    $[76] = t21;
  } else {
    t21 = $[76];
  }
  const containerProps = t21;
  let t22;
  if ($[77] !== _PrivateItemWrapper || $[78] !== forwardedRef || $[79] !== listSemantics || $[80] !== menuItemProps || $[81] !== props) {
    t22 = _PrivateItemWrapper ? menuItemProps : !listSemantics && {
      ...menuItemProps,
      ...props,
      ref: forwardedRef
    };
    $[77] = _PrivateItemWrapper;
    $[78] = forwardedRef;
    $[79] = listSemantics;
    $[80] = menuItemProps;
    $[81] = props;
    $[82] = t22;
  } else {
    t22 = $[82];
  }
  const wrapperProps = t22;
  const t23 = Boolean(inactiveText);
  const t24 = buttonSemantics ? setTruncatedText : undefined;
  let t25;
  if ($[83] !== blockDescriptionId || $[84] !== disabled || $[85] !== inlineDescriptionId || $[86] !== size || $[87] !== t23 || $[88] !== t24 || $[89] !== trailingVisualId || $[90] !== variant) {
    t25 = {
      variant,
      size,
      disabled,
      inactive: t23,
      inlineDescriptionId,
      blockDescriptionId,
      trailingVisualId,
      setTruncatedText: t24
    };
    $[83] = blockDescriptionId;
    $[84] = disabled;
    $[85] = inlineDescriptionId;
    $[86] = size;
    $[87] = t23;
    $[88] = t24;
    $[89] = trailingVisualId;
    $[90] = variant;
    $[91] = t25;
  } else {
    t25 = $[91];
  }
  const itemContextValue = t25;
  const t26 = listSemantics ? forwardedRef : null;
  const t27 = variant === "danger" ? variant : undefined;
  const t28 = active ? true : undefined;
  const t29 = inactiveText ? true : undefined;
  const t30 = disabled ? true : undefined;
  const t31 = slots.subItem ? true : undefined;
  const t32 = slots.description ? true : false;
  let t33;
  if ($[92] !== className) {
    t33 = clsx(classes.ActionListItem, className);
    $[92] = className;
    $[93] = t33;
  } else {
    t33 = $[93];
  }
  let t34;
  let t35;
  if ($[94] === Symbol.for("react.memo_cache_sentinel")) {
    t34 = {};
    t35 = /*#__PURE__*/jsx("span", {
      className: classes.Spacer
    });
    $[94] = t34;
    $[95] = t35;
  } else {
    t34 = $[94];
    t35 = $[95];
  }
  let t36;
  if ($[96] !== selected) {
    t36 = /*#__PURE__*/jsx(Selection, {
      selected: selected,
      className: classes.LeadingAction
    });
    $[96] = selected;
    $[97] = t36;
  } else {
    t36 = $[97];
  }
  const t37 = showInactiveIndicator ? inactiveText : undefined;
  const t38 = Boolean(slots.leadingVisual);
  let t39;
  if ($[98] !== labelId || $[99] !== loading || $[100] !== slots.leadingVisual || $[101] !== t37 || $[102] !== t38) {
    t39 = /*#__PURE__*/jsx(VisualOrIndicator, {
      inactiveText: t37,
      itemHasLeadingVisual: t38,
      labelId: labelId,
      loading: loading,
      position: "leading",
      children: slots.leadingVisual
    });
    $[98] = labelId;
    $[99] = loading;
    $[100] = slots.leadingVisual;
    $[101] = t37;
    $[102] = t38;
    $[103] = t39;
  } else {
    t39 = $[103];
  }
  const t40 = !!slots.description;
  let t41;
  if ($[104] !== inactive || $[105] !== loading) {
    t41 = loading === true && !inactive && /*#__PURE__*/jsx(VisuallyHidden, {
      children: "Loading"
    });
    $[104] = inactive;
    $[105] = loading;
    $[106] = t41;
  } else {
    t41 = $[106];
  }
  let t42;
  if ($[107] !== childrenWithoutSlots || $[108] !== labelId || $[109] !== t41) {
    t42 = /*#__PURE__*/jsxs("span", {
      id: labelId,
      className: classes.ItemLabel,
      children: [childrenWithoutSlots, t41]
    });
    $[107] = childrenWithoutSlots;
    $[108] = labelId;
    $[109] = t41;
    $[110] = t42;
  } else {
    t42 = $[110];
  }
  let t43;
  if ($[111] !== descriptionVariant || $[112] !== slots.description || $[113] !== t40 || $[114] !== t42) {
    t43 = /*#__PURE__*/jsxs(ConditionalWrapper, {
      if: t40,
      className: classes.ItemDescriptionWrap,
      "data-description-variant": descriptionVariant,
      children: [t42, slots.description]
    });
    $[111] = descriptionVariant;
    $[112] = slots.description;
    $[113] = t40;
    $[114] = t42;
    $[115] = t43;
  } else {
    t43 = $[115];
  }
  const t44 = showInactiveIndicator ? inactiveText : undefined;
  const t45 = Boolean(slots.leadingVisual);
  let t46;
  if ($[116] !== labelId || $[117] !== loading || $[118] !== t44 || $[119] !== t45 || $[120] !== trailingVisual) {
    t46 = /*#__PURE__*/jsx(VisualOrIndicator, {
      inactiveText: t44,
      itemHasLeadingVisual: t45,
      labelId: labelId,
      loading: loading,
      position: "trailing",
      children: trailingVisual
    });
    $[116] = labelId;
    $[117] = loading;
    $[118] = t44;
    $[119] = t45;
    $[120] = trailingVisual;
    $[121] = t46;
  } else {
    t46 = $[121];
  }
  let t47;
  if ($[122] !== inactiveText || $[123] !== inactiveWarningId || $[124] !== showInactiveIndicator) {
    t47 = !showInactiveIndicator && inactiveText ? /*#__PURE__*/jsx("span", {
      className: classes.InactiveWarning,
      id: inactiveWarningId,
      children: inactiveText
    }) : null;
    $[122] = inactiveText;
    $[123] = inactiveWarningId;
    $[124] = showInactiveIndicator;
    $[125] = t47;
  } else {
    t47 = $[125];
  }
  let t48;
  if ($[126] !== t43 || $[127] !== t46 || $[128] !== t47) {
    t48 = /*#__PURE__*/jsxs("span", {
      className: classes.ActionListSubContent,
      "data-component": "ActionList.Item--DividerContainer",
      children: [t43, t46, t47]
    });
    $[126] = t43;
    $[127] = t46;
    $[128] = t47;
    $[129] = t48;
  } else {
    t48 = $[129];
  }
  let t49;
  if ($[130] !== t36 || $[131] !== t39 || $[132] !== t48) {
    t49 = /*#__PURE__*/jsxs(TooltipContext.Provider, {
      value: t34,
      children: [t35, t36, t39, t48]
    });
    $[130] = t36;
    $[131] = t39;
    $[132] = t48;
    $[133] = t49;
  } else {
    t49 = $[133];
  }
  let t50;
  if ($[134] !== ItemWrapper || $[135] !== forwardedRef || $[136] !== size || $[137] !== t49 || $[138] !== wrapperProps) {
    t50 = /*#__PURE__*/jsx(ItemWrapper, {
      ...wrapperProps,
      className: classes.ActionListContent,
      "data-size": size,
      ref: forwardedRef,
      children: t49
    });
    $[134] = ItemWrapper;
    $[135] = forwardedRef;
    $[136] = size;
    $[137] = t49;
    $[138] = wrapperProps;
    $[139] = t50;
  } else {
    t50 = $[139];
  }
  let t51;
  if ($[140] !== buttonSemantics || $[141] !== forwardedRef || $[142] !== t50 || $[143] !== truncatedText) {
    t51 = /*#__PURE__*/jsx(ConditionalTooltip, {
      ref: forwardedRef,
      text: truncatedText,
      enabled: buttonSemantics,
      children: t50
    });
    $[140] = buttonSemantics;
    $[141] = forwardedRef;
    $[142] = t50;
    $[143] = truncatedText;
    $[144] = t51;
  } else {
    t51 = $[144];
  }
  const t52 = !inactive && !loading && !menuContext && Boolean(slots.trailingAction) && slots.trailingAction;
  let t53;
  if ($[145] !== containerProps || $[146] !== slots.subItem || $[147] !== t26 || $[148] !== t27 || $[149] !== t28 || $[150] !== t29 || $[151] !== t30 || $[152] !== t31 || $[153] !== t32 || $[154] !== t33 || $[155] !== t51 || $[156] !== t52) {
    t53 = /*#__PURE__*/jsxs("li", {
      ...containerProps,
      ref: t26,
      "data-variant": t27,
      "data-active": t28,
      "data-inactive": t29,
      "data-is-disabled": t30,
      "data-has-subitem": t31,
      "data-has-description": t32,
      className: t33,
      children: [t51, t52, slots.subItem]
    });
    $[145] = containerProps;
    $[146] = slots.subItem;
    $[147] = t26;
    $[148] = t27;
    $[149] = t28;
    $[150] = t29;
    $[151] = t30;
    $[152] = t31;
    $[153] = t32;
    $[154] = t33;
    $[155] = t51;
    $[156] = t52;
    $[157] = t53;
  } else {
    t53 = $[157];
  }
  let t54;
  if ($[158] !== itemContextValue || $[159] !== t53) {
    t54 = /*#__PURE__*/jsx(ItemContext.Provider, {
      value: itemContextValue,
      children: t53
    });
    $[158] = itemContextValue;
    $[159] = t53;
    $[160] = t54;
  } else {
    t54 = $[160];
  }
  return t54;
};
const Item = Object.assign(fixedForwardRef(UnwrappedItem), {
  displayName: 'ActionList.Item',
  __SLOT__: Symbol('ActionList.Item')
});

export { Item, SubItem };
